### PR TITLE
Parity-db `Change` missing implementation. 

### PR DESCRIPTION
**Pull request by cheme**
This pr is a minimal implementation of Change::Store, Change::Reference, Change::Release, in order for parity db to support storage chain (those operator are only in this case).

The implementation is really not optimal: when using reference, we should not fetch the value, a parity-db specific instruction should be use (only need to update the internal storage rc), with current `commit` api adding a new instruction requires a bit of rework so having this quick PR first make sense to me.

In case column do not use rc, I choose to not use the default implementation from kvdb as it indicates that something should be change.

Polkadot companion: https://github.com/paritytech/polkadot/pull/5472

### CODE REVIEW

**Review by arkpar — APPROVED** [review]

File: client/db/src/parity_db.rs [hunk 1]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
+			Change::Store(col, key, value) =>
+				if ref_counted_column(col) {
+					(col as u8, key.as_ref().to_vec(), Some(value))
+				} else {
+					unimplemented!()
+				},
+			Change::Reference(col, key) =>
+				if ref_counted_column(col) {
+					// FIXME accessing value is not strictly needed, optimize this in parity-db.
+					let value = <Self as Database<H>>::get(self, col, key.as_ref());
```
[the remark this answers is not in the document]
Comment (cheme) [reply client/db/src/parity_db.rs:107]: yes, I also realize this morning that we always read the same counter value.
I did not fix it, as for storage chain it is not an issue and I do not know of other usage of these Change variants.

File: client/db/src/parity_db.rs [hunk 2]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
 	fn commit(&self, transaction: Transaction<H>) -> Result<(), DatabaseError> {
 		handle_err(self.0.commit(transaction.0.into_iter().map(|change| match change {
 			Change::Set(col, key, value) => (col as u8, key, Some(value)),
 			Change::Remove(col, key) => (col as u8, key, None),
-			_ => unimplemented!(),
+			Change::Store(col, key, value) =>
+				if ref_counted_column(col) {
+					(col as u8, key.as_ref().to_vec(), Some(value))
+				} else {
+					unimplemented!()
```
Comment (arkpar) [inline client/db/src/parity_db.rs:102]:
```suggestion
					expect("Change::Store is only used for ref counted columns")
```

File: client/db/src/parity_db.rs [hunk 3]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
+				},
+			Change::Reference(col, key) =>
+				if ref_counted_column(col) {
+					// FIXME accessing value is not strictly needed, optimize this in parity-db.
+					let value = <Self as Database<H>>::get(self, col, key.as_ref());
+					(col as u8, key.as_ref().to_vec(), value)
+				} else {
+					unimplemented!()
```
Comment (arkpar) [inline client/db/src/parity_db.rs:110]:
```suggestion
					panic!("Change::Reference is only used for ref counted columns")
```

File: client/db/src/parity_db.rs [hunk 4]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
+				},
+			Change::Release(col, key) =>
+				if ref_counted_column(col) {
+					(col as u8, key.as_ref().to_vec(), None)
+				} else {
+					unimplemented!()
```
Comment (arkpar) [inline client/db/src/parity_db.rs:116]:
```suggestion
					panic!("Change::Release is only used for ref counted columns")
```

**Comment by cheme** [comment]
I totally lost track of this PR, still needed.

File: client/db/src/parity_db.rs [hunk 5]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
 	fn commit(&self, transaction: Transaction<H>) -> Result<(), DatabaseError> {
 		handle_err(self.0.commit(transaction.0.into_iter().map(|change| match change {
 			Change::Set(col, key, value) => (col as u8, key, Some(value)),
 			Change::Remove(col, key) => (col as u8, key, None),
-			_ => unimplemented!(),
+			Change::Store(col, key, value) =>
+				if ref_counted_column(col) {
+					(col as u8, key.as_ref().to_vec(), Some(value))
+				} else {
+					panic!("Change::Store is only used for ref counted columns")
```
Comment (bkchr) [inline client/db/src/parity_db.rs:102]: If that is the case, why do we still need the if around it?

Comment (bkchr) [inline client/db/src/parity_db.rs:102]:
```suggestion
					unreachable!("Change::Store is only used for ref counted columns")
```

File: client/db/src/parity_db.rs [hunk 6]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
+					let value = <Self as Database<H>>::get(self, col, key.as_ref());
+					(col as u8, key.as_ref().to_vec(), value)
+				} else {
+					panic!("Change::Reference is only used for ref counted columns")
+				},
+			Change::Release(col, key) =>
+				if ref_counted_column(col) {
+					(col as u8, key.as_ref().to_vec(), None)
+				} else {
+					panic!("Change::Release is only used for ref counted columns")
```
Comment (bkchr) [inline client/db/src/parity_db.rs:116]:
```suggestion
					unreachable!("Change::Release is only used for ref counted columns")
```

File: client/db/src/parity_db.rs [hunk 7]
```diff
@@ -86,12 +86,35 @@ pub fn open<H: Clone + AsRef<[u8]>>(
...
+				} else {
+					panic!("Change::Store is only used for ref counted columns")
+				},
+			Change::Reference(col, key) =>
+				if ref_counted_column(col) {
+					// FIXME accessing value is not strictly needed, optimize this in parity-db.
+					let value = <Self as Database<H>>::get(self, col, key.as_ref());
+					(col as u8, key.as_ref().to_vec(), value)
+				} else {
+					panic!("Change::Reference is only used for ref counted columns")
```
Comment (bkchr) [inline client/db/src/parity_db.rs:110]:
```suggestion
					unreachable!("Change::Reference is only used for ref counted columns")
```

**Review by bkchr — APPROVED** [review]

**Comment by cheme** [comment]
@bkchr , @arkpar , I did change the code a bit by returning error with the `commit` method of the adapter (previously error on commit was panicking).
May not be a great idea, but I think it can be a good direction. I can revert though (ensuring no code regression).

**Review by bkchr — APPROVED** [review]

**Comment by arkpar** [comment]
bot merge

**Comment by cheme** [comment]
bot merge

**Pull request merged**